### PR TITLE
Adds a new tool to replace gen_domain in CIME

### PR DESCRIPTION
**Pull request by whannah1**
"generate_domain_files_E3SM.py" is a new tool to replace the fortran
gen_domain.F90 tool that exists in CIME. The old version was problematic
to use because the build configuration was almost always broken when a
user went to build it. The new python version works efficiently due to
the numba package, which is included in the E3SM unified environment.

[BFB]

### CODE REVIEW

**Review by singhbalwinder — APPROVED** [review]
Thanks for rewriting this tool! I left some very minor comments.

File: tools/generate_domain_files/generate_domain_files_E3SM.py
Comment (whannah1) [reply tools/generate_domain_files/generate_domain_files_E3SM.py:124]: I don't really like that convention since it feels so arbitrary, but I'm fine applying it here.

Comment (whannah1) [reply tools/generate_domain_files/generate_domain_files_E3SM.py:124]: Actually, I just realized we never actually use this variable, in both the old fortran tool and my code. It's written to the file, but I don't think the model actually uses it, so I'm going to delete all mentions of it.

File: tools/generate_domain_files/generate_domain_files_E3SM.py
Comment (whannah1) [reply tools/generate_domain_files/generate_domain_files_E3SM.py:339]: I actually did this at the start, but it felt unnecessary. I do like the practice in general though, so I'll refactor it.

Comment (whannah1) [reply tools/generate_domain_files/generate_domain_files_E3SM.py]: ok, I refactored it to have a "main" routine.

**Comment by whannah1** [comment]
I think this PR is ready for a thorough review. 

I've added documentation for this tool and a new "tools" section of the E3SM docs. I expect other tool documentaiton will show up there soon. 

Additionally, I've also started working on recreating the guide for "Running E3SM on New Grids" in the new E3SM docs repo. So there will naturally be some redundant documentation for this tool, but I think that's ok.

File: mkdocs.yaml
Comment (whannah1) [reply mkdocs.yaml:7]: This was a bit puzzling because I noticed at the top level we have "mkdocs.yaml", but elsewhere we have "mkdocs.yml". I guess the top-level one should be changed?

Comment (whannah1) [reply mkdocs.yaml:7]: done

File: tools/generate_domain_files/docs/index.md
Comment (whannah1) [reply tools/generate_domain_files/docs/index.md:14]: I didn't realize "shell" was an option, I like that better.

Comment (whannah1) [reply tools/generate_domain_files/docs/index.md:14]: done

File: tools/generate_domain_files/mkdocs.yaml
Comment (whannah1) [reply tools/generate_domain_files/mkdocs.yaml:1]: Yea, I'm a bit unsure about the best strategy here as well. Initially I was envisioning a more complicated hierarchy, but then Ben Hillman pointed out that I was committing the sin of premature optimization. So I think it's best to see how the documentation evolves and what new tools get added before we make any decisions about how to reorganize things.

**Review by bishtgautam — APPROVED** [review]

**Comment by singhbalwinder** [comment]
Hi @jonbob and @mahf708: Do you guys have any further comments on this PR before I merge it? Thanks!

**Comment by jonbob** [comment]
@singhbalwinder - I was going to test it and compare output from the Fortran tool quickly, but I'll do that today and post results here

**Comment by jonbob** [comment]
@whannah1 -- it failed for me trying to regenerate a domain file from one of our recent ice/ocn meshes:
````
python generate_domain_files_E3SM.py -m /lcrc/group/acme/ac.jwolfe/v3-grids/RRS6to18E3r6/map_RRS6to18E3r6_to_TL319_traave.20240403.nc -o RRS6to18E3r6 -l TL319

  Input files and parameter values:
    map_file        : /lcrc/group/acme/ac.jwolfe/v3-grids/RRS6to18E3r6/map_RRS6to18E3r6_to_TL319_traave.20240403.nc
    lnd_grid        : TL319
    ocn_grid        : RRS6to18E3r6
    fminval         : 0
    fmaxval         : 1
    set_omask       : False
  

  Grid information from map file:
    domain_a file        : ocean.RRS6to18E3r6.scrip.20240402.nc
    domain_b file        : TL319.151007.nc
    ocn_grid_file        : ocean.RRS6to18E3r6.scrip.20240402.nc
    atm_grid_file        : TL319.151007.nc
    ocn grid size   (n_a): 4016094
    atm grid size   (n_b): 204800
    sparse mat size (n_s): 5992418
  
Traceback (most recent call last):
  File "/lcrc/group/acme/ac.jwolfe/e3sm_PR6347/tools/generate_domain_files/generate_domain_files_E3SM.py", line 360, in <module>
    main()
  File "/lcrc/group/acme/ac.jwolfe/e3sm_PR6347/tools/generate_domain_files/generate_domain_files_E3SM.py", line 211, in main
    add_metadata(ds_out)
  File "/lcrc/group/acme/ac.jwolfe/e3sm_PR6347/tools/generate_domain_files/generate_domain_files_E3SM.py", line 301, in add_metadata
    ds_out['xc'] = ds_out['xc'].assign_attrs({'long_name':'longitude of grid cell center'})
NameError: name 'ds_out' is not defined
````

Of course it's highly likely to be user error

**Comment by rljacob** [comment]
I agree with @mahf708.  Move the docs directory to just under tools.

**Comment by whannah1** [comment]
After discussing how to restructure the tools docs with @rljacob we didn't like the alternative, so we decided to leave it as is. 

I also fixed the bug that @jonbob found.

**Comment by whannah1** [comment]
@singhbalwinder looks like I have another problem with the output domain files that I need to investigate, so we need to hold off on merging this.

**Review by jonbob — APPROVED** [review]
Approved based on testing, which showed BFB results comparing a land domain file made with this tool vs one made with the Fortran version

**Comment by whannah1** [comment]
@mahf708 I want to hold off on the testing idea mainly because I'm not sure how to write a coherent test. We could set it up to just go through motions and prove that it can produce a set of files, but I'm not sure how to test the output for any sort of correctness. So let's hold off. 

@singhbalwinder The issue that Jon identified is now fixed, so this PR is ready to be merged.

**Comment by whannah1** [comment]
As one last test I ran a quick test using F2010 on the ne30pg2_oECv3 grid replacing the default domain files with ones produced by the new tool and it ran without issue. However, I didn't do a baseline comparison.

**Comment by singhbalwinder** [comment]
Merged to next

**Pull request merged**
